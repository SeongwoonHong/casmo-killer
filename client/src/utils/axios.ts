import axios from 'axios';

axios.interceptors.request.use((config) => {
  console.log(config);
  return config;
});

function setTokenToHeader(header, csrfToken) {
  axios.defaults.headers.common[header] = csrfToken;
}

export {
  axios,
  setTokenToHeader,
};
