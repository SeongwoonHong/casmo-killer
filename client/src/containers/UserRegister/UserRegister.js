import React, { Component } from 'react';
import axios from 'axios';
import { Prompt } from 'react-router-dom';

import * as storage from '@sharedUtils/storage';
import {
  validateImg,
  validateDisplayName,
  validatePassword
} from '@sharedUtils/inputValidators';

import FormMessage from '@sharedComponents/FormMessage';

import UserPageContainer from '../../components/UserPageContainer';
import UserInputField from '../../components/UserInputField';

import './UserRegister.scss';

class Register extends Component {

  constructor(props) {

    super(props);

    this.state = {
      isMounted: false,
      isLoading: false,
      displayName: {
        value: props.auth.displayName || '',
        message: ''
      },
      password: {
        value: '',
        message: ''
      },
      avatar: {
        value: props.auth.avatar || '',
        message: ''
      },
      errorMsg: ''
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onImageUpload = this.onImageUpload.bind(this);
    this.onSubmitHandler = this.onSubmitHandler.bind(this);
    this.submitRegister = this.submitRegister.bind(this);
    this.verifyToken = this.verifyToken.bind(this);

  }

  componentDidMount() {

    const {
      history, auth, match, setErrorState
    } = this.props;

    if (match.params.token) {

      this.verifyToken(match.params.token);

    } else if (!auth.email) {

      setErrorState({
        errorTitle: 'The registration link is invalid.',
        errorMsg: 'Please submit your email again to complete the registration.'
      });

      history.push('/error');

    }

  }

  onChangeHandler(e) {
    this.setState({
      [e.name]: {
        ...this.state[e.name],
        value: e.value
      }
    });
  }

  async onSubmitHandler() {

    const resetState = {};

    // remove any previous validation messages
    Object.keys(this.state).forEach((key) => {
      resetState[key] = this.state[key];
      if (Object.prototype.hasOwnProperty.call(this.state[key], 'message')) {
        resetState[key].message = '';
      }
    });

    // empty the messages and set the loading status to true
    this.setState(Object.assign({}, resetState, {
      isLoading: true
    }));

    const { displayName, password } = this.state;

    const errorMsg = {
      forDisplayName: '',
      forPassword: ''
    };

    // validate displayName and register validation messages
    // if there are any
    errorMsg.forDisplayName = await validateDisplayName(displayName.value);

    this.setState({
      displayName: {
        ...this.state.displayName,
        message: errorMsg.forDisplayName
      }
    });

    // validate the password only when it's local signup
    if (this.props.auth.strategy === 'local') {

      errorMsg.forPassword = await validatePassword(password.value);

      this.setState({
        password: {
          ...this.state.password,
          message: errorMsg.forPassword
        }
      });

    }

    if (
      errorMsg.forDisplayName.length === 0 &&
      errorMsg.forPassword.length === 0
    ) {
      this.submitRegister();
    } else {
      this.setState({ isLoading: false });
    }

  }

  // TODO: move image upload to its own component
  onImageUpload(e) {

    this.setState({ isLoading: true });

    const reader = new FileReader();
    const image = e.target.files[0];

    let message = '';

    reader.onloadend = () => {

      if (image.size > 5000000) {
        message = 'File is too big.';
      } else if (!validateImg(reader.result)) {
        message = 'File format is not supported.';
      }

      this.setState({
        avatar: {
          value: message.length > 0
            ? this.state.avatar.value
            : reader.result,
          message
        },
        isLoading: false
      });

    };

    reader.readAsDataURL(image);

  }

  async submitRegister() {

    const {
      strategy, email, socialId, socialToken
    } = this.props.auth;

    const {
      displayName, avatar, password
    } = this.state;

    const isLocal = strategy === 'local';

    const userData = new FormData();

    userData.append('strategy', strategy);
    userData.append('email', email);
    userData.append('displayName', displayName.value);

    if (avatar.value.length > 0) {
      userData.append('avatar', avatar.value);
    }

    if (isLocal) {
      userData.append('password', password.value);
    } else {
      userData.append('socialId', socialId);
      userData.append('socialToken', socialToken);
    }

    try {

      const apiUrl = `/api/auth/register/${isLocal ? 'local' : 'social'}`;
      const { data } = await axios.post(apiUrl, userData);

      if (data && data.user) {

        const user = await storage.set('ckUser', data.user);

        await this.props.loginSuccess(user, true);

        const { isLoggedIn } = this.props.user;

        if (isLoggedIn) {
          this.props.history.replace('/');
        }

      } else {

        this.setState({
          errorMsg: 'Failed to communicate with the server.'
        });

      }

    } catch (error) {

      console.error(error);
      console.error(error.response.data.error);
      this.setState({
        errorMsg: error.response && error.response.data.message
      });

    }

    this.setState({ isLoading: false });

  }

  async verifyToken(token) {

    const {
      history, setUserForRegistration, setErrorState
    } = this.props;

    try {

      const { data } = await axios.get(`/api/auth/verify/token/register/${token}`);

      if (data && data.email) {

        setUserForRegistration({
          strategy: 'local',
          email: data.email
        });

        history.replace('/user/register');

      } else {

        this.setState({
          errorMsg: 'Failed to communicate with the server.'
        });

      }

    } catch (error) {

      console.error(error);

      setErrorState({
        errorTitle: 'The registration link is invalid.',
        errorMsg: error.response && error.response.data.message
      });

      history.push('/error');

    }

    this.setState({ isMounted: true });

  }

  render() {

    const {
      isMounted,
      isLoading,
      displayName,
      password,
      avatar,
      errorMsg
    } = this.state;

    const { auth } = this.props;

    const avatarPreview = (avatarState) => {
      if (avatarState.value) {
        return (
          <img
            className="circle avatar-img"
            alt="user-avatar"
            src={ avatarState.value }
          />
        );
      }
      return <span>No Image</span>;
    };

    return (
      <UserPageContainer
        className="Register"
        title="User Registration"
        icon="person_add"
        isLoading={ isLoading }
        formTitle="User Information"
        formMsg="Please fill in the following fields to complete your registration."
        onSubmit={ this.onSubmitHandler }>

        <Prompt
          when={ isMounted && !isLoading }
          message="Are you fucking sure? Your information will be lost." />

        <FormMessage message={ errorMsg } />

        <UserInputField
          name="email"
          onChange={ this.onChangeHandler }
          value={ auth.email || '' }
          disabled={ !!auth.email } />

        <FormMessage message={ displayName.message } />
        <UserInputField
          name="displayName"
          onChange={ this.onChangeHandler }
          value={ displayName.value } />

        <FormMessage message={ password.message } />
        <UserInputField
          isVisible={ auth.strategy === 'local' }
          name="password"
          onChange={ this.onChangeHandler }
          value={ password.value } />

        <FormMessage message={ avatar.message } />
        <div className="user-form-fields">
          <label>Profile Picture</label>
          <div className="avatar-preview">
            <div className="avatar-wrapper">
              { avatarPreview(avatar) }
            </div>
            <input
              type="file"
              accept="image/*"
              id="profilePicture"
              onChange={ this.onImageUpload } />
            <label htmlFor="profilePicture">
              <span className="user-form-button">
                Upload New Picture
              </span>
              <span>Max 5mb, JPG, or PNG</span>
            </label>
          </div>
        </div>

      </UserPageContainer>
    );

  }

}

export default Register;
