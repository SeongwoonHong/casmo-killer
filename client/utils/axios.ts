import axios from 'axios';

const baseURL = 'http://localhost:9000';

axios.defaults.baseURL = baseURL + '/api';

export { axios };
