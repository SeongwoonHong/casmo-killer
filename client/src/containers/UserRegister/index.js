import { connect } from 'react-redux';
import * as actions from '@actions';

import UserRegister from './UserRegister';

function mapStateToProps(state) {
  return {
    auth: state.auth,
    user: state.user.user
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setUserForRegistration: payload => dispatch(actions.setUserForRegistration(payload)),
    loginSuccess: (payload, shouldReset) => dispatch(actions.loginSuccess(payload, shouldReset)),
    clearRedirectUrl: () => dispatch(actions.clearRedirectUrl()),
    setErrorState: error => dispatch(actions.setErrorState(error))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserRegister);
