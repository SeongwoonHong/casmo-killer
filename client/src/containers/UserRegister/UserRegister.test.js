/* eslint react/jsx-boolean-value: 0 */
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import axios from 'axios';

import UserRegister from './UserRegister';

describe('<UserRegister />', () => {

  let component;
  let mockGet;
  let mockPost;

  const auth = {
    strategy: 'facebook',
    email: 'ssinsoo@gmail.com',
    password: null,
    displayName: 'LyndonOhhyungChun',
    avatar: 'https://lookaside.facebook.com/platform/profilepic/?asid=1527248567354389&height=50&width=50&ext=1522858773&hash=AeTOk-4ueDQXWd1T',
    socialId: '1527248567354389',
    socialToken: 'EAAB86Ub9xDcBAFJCETkRP2KeSR2ZBEVSf02hXXpTLuDhSRGNTXMjroXJwu6UNDidVFFpElN7DUoC5ifhVJiSAWNQ3kwAxr5rNKBFZC9SvCZBg4mcKc0QZBzYGrNru5vwUEHo2ly71zAcyXLy1TIDuc0hdhTTWXfACLyPSuP1IW6gr1MT8hI5bEIwS91SDLdCipHbBe7DSAZDZD',
    redirectUrl: '/',
    verificationToken: null,
    message: ''
  };

  beforeEach(() => {

    component = shallow(
      <UserRegister
        auth={ auth }
        match={ { params: { token: auth.token } } } />
    );

    mockGet = sinon.stub(axios, 'get');
    mockPost = sinon.stub(axios, 'post');

  });

  afterEach(() => {
    mockGet.restore();
    mockPost.restore();
  });

  it('should render the component consistently', () => {
    expect(component).toMatchSnapshot();
  });

  it('should construct the component state with correct props', () => {
    expect(component.state()).toEqual({
      isMounted: true,
      isLoading: false,
      displayName: {
        value: auth.displayName,
        message: ''
      },
      password: {
        value: '',
        message: ''
      },
      avatar: {
        value: auth.avatar,
        message: ''
      },
      errorMsg: ''
    });
  });

  it('should submit the form and empty remove messages', async (done) => {

    const loginSuccess = sinon.spy();
    const replace = sinon.spy();

    const userData = { data: { isDuplicate: false } };
    const registrationData = { data: { user: { isLoggedIn: true } } };

    const promise = Promise.resolve(userData);
    const registrationPromise = Promise.resolve(registrationData);

    mockGet.callsFake(() => promise);
    mockPost.callsFake(() => registrationPromise);

    const wrapper = shallow(
      <UserRegister
        auth={ auth }
        user={ registrationData.data.user }
        loginSuccess={ loginSuccess }
        history={ { replace } }
        match={ { params: { token: auth.socialToken } } } />
    );

    wrapper.instance().onSubmitHandler();

    await promise;
    await registrationPromise;

    expect(component.state('displayName').message).toEqual('');
    expect(component.state('password').message).toEqual('');
    expect(component.state('avatar').message).toEqual('');

    // TODO: figure out a way to reorder the test
    setImmediate(() => {
      expect(loginSuccess.calledWith(registrationData.data.user, true)).toEqual(true);
      expect(replace.calledWith('/')).toEqual(true);
      done();
    });

  });

  it('should verify the token for registration', async () => {

    const setUserForRegistration = sinon.spy();
    const replace = sinon.spy();

    const userData = { data: { email: 'ckboard@gmail.com' } };
    const promise = Promise.resolve(userData);

    mockGet.callsFake(() => promise);

    shallow(
      <UserRegister
        auth={ auth }
        setUserForRegistration={ setUserForRegistration }
        history={ { replace } }
        match={ { params: { token: auth.socialToken } } } />
    );

    await promise;

    expect(setUserForRegistration.calledWith({
      strategy: 'local',
      email: userData.data.email
    })).toEqual(true);
    expect(replace.calledWith('/user/register')).toEqual(true);

  });

  it('should redirect to error page when there is no token and email', () => {

    const setErrorState = sinon.spy();
    const push = sinon.spy();

    shallow(
      <UserRegister
        auth={ {} }
        setErrorState={ setErrorState }
        history={ { push } }
        match={ { params: { token: null } } } />
    );

    expect(setErrorState.calledWith({
      errorTitle: 'The registration link is invalid.',
      errorMsg: 'Please submit your email again to complete the registration.'
    })).toEqual(true);
    expect(push.calledWith('/error')).toEqual(true);

  });

});
