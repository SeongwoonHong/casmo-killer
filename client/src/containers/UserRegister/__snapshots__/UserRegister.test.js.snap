// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserRegister /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Register
    auth={
        Object {
            "avatar": "https://lookaside.facebook.com/platform/profilepic/?asid=1527248567354389&height=50&width=50&ext=1522858773&hash=AeTOk-4ueDQXWd1T",
            "displayName": "LyndonOhhyungChun",
            "email": "ssinsoo@gmail.com",
            "message": "",
            "password": null,
            "redirectUrl": "/",
            "socialId": "1527248567354389",
            "socialToken": "EAAB86Ub9xDcBAFJCETkRP2KeSR2ZBEVSf02hXXpTLuDhSRGNTXMjroXJwu6UNDidVFFpElN7DUoC5ifhVJiSAWNQ3kwAxr5rNKBFZC9SvCZBg4mcKc0QZBzYGrNru5vwUEHo2ly71zAcyXLy1TIDuc0hdhTTWXfACLyPSuP1IW6gr1MT8hI5bEIwS91SDLdCipHbBe7DSAZDZD",
            "strategy": "facebook",
            "verificationToken": null,
          }
    }
    match={
        Object {
            "params": Object {
              "token": undefined,
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Prompt
          message="Are you fucking sure? Your information will be lost."
          when={false}
/>,
        <FormMessage
          className=""
          message=""
          type="error"
/>,
        <UserInputField
          disabled={true}
          isVisible={true}
          message=""
          name="email"
          onChange={[Function]}
          placeholder=""
          title=""
          type=""
          value="ssinsoo@gmail.com"
/>,
        <FormMessage
          className=""
          message=""
          type="error"
/>,
        <UserInputField
          disabled={false}
          isVisible={true}
          message=""
          name="displayName"
          onChange={[Function]}
          placeholder=""
          title=""
          type=""
          value="LyndonOhhyungChun"
/>,
        <FormMessage
          className=""
          message=""
          type="error"
/>,
        <UserInputField
          disabled={false}
          isVisible={false}
          message=""
          name="password"
          onChange={[Function]}
          placeholder=""
          title=""
          type=""
          value=""
/>,
        <FormMessage
          className=""
          message=""
          type="error"
/>,
        <AvatarUploader
          avatar={
                    Object {
                              "message": "",
                              "value": "https://lookaside.facebook.com/platform/profilepic/?asid=1527248567354389&height=50&width=50&ext=1522858773&hash=AeTOk-4ueDQXWd1T",
                            }
          }
          className="User__form__fields"
          onChange={[Function]}
/>,
      ],
      "className": "Register",
      "disabled": false,
      "formMsg": "Please fill in the following fields to complete your registration.",
      "formTitle": "User Information",
      "icon": "person_add",
      "isLoading": false,
      "onSubmit": [Function],
      "title": "User Registration",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "message": "Are you fucking sure? Your information will be lost.",
          "when": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "",
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": true,
          "isVisible": true,
          "message": "",
          "name": "email",
          "onChange": [Function],
          "placeholder": "",
          "title": "",
          "type": "",
          "value": "ssinsoo@gmail.com",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "",
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": false,
          "isVisible": true,
          "message": "",
          "name": "displayName",
          "onChange": [Function],
          "placeholder": "",
          "title": "",
          "type": "",
          "value": "LyndonOhhyungChun",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "",
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": false,
          "isVisible": false,
          "message": "",
          "name": "password",
          "onChange": [Function],
          "placeholder": "",
          "title": "",
          "type": "",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "",
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "avatar": Object {
            "message": "",
            "value": "https://lookaside.facebook.com/platform/profilepic/?asid=1527248567354389&height=50&width=50&ext=1522858773&hash=AeTOk-4ueDQXWd1T",
          },
          "className": "User__form__fields",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Prompt
            message="Are you fucking sure? Your information will be lost."
            when={false}
/>,
          <FormMessage
            className=""
            message=""
            type="error"
/>,
          <UserInputField
            disabled={true}
            isVisible={true}
            message=""
            name="email"
            onChange={[Function]}
            placeholder=""
            title=""
            type=""
            value="ssinsoo@gmail.com"
/>,
          <FormMessage
            className=""
            message=""
            type="error"
/>,
          <UserInputField
            disabled={false}
            isVisible={true}
            message=""
            name="displayName"
            onChange={[Function]}
            placeholder=""
            title=""
            type=""
            value="LyndonOhhyungChun"
/>,
          <FormMessage
            className=""
            message=""
            type="error"
/>,
          <UserInputField
            disabled={false}
            isVisible={false}
            message=""
            name="password"
            onChange={[Function]}
            placeholder=""
            title=""
            type=""
            value=""
/>,
          <FormMessage
            className=""
            message=""
            type="error"
/>,
          <AvatarUploader
            avatar={
                        Object {
                                    "message": "",
                                    "value": "https://lookaside.facebook.com/platform/profilepic/?asid=1527248567354389&height=50&width=50&ext=1522858773&hash=AeTOk-4ueDQXWd1T",
                                  }
            }
            className="User__form__fields"
            onChange={[Function]}
/>,
        ],
        "className": "Register",
        "disabled": false,
        "formMsg": "Please fill in the following fields to complete your registration.",
        "formTitle": "User Information",
        "icon": "person_add",
        "isLoading": false,
        "onSubmit": [Function],
        "title": "User Registration",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "message": "Are you fucking sure? Your information will be lost.",
            "when": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": true,
            "isVisible": true,
            "message": "",
            "name": "email",
            "onChange": [Function],
            "placeholder": "",
            "title": "",
            "type": "",
            "value": "ssinsoo@gmail.com",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": false,
            "isVisible": true,
            "message": "",
            "name": "displayName",
            "onChange": [Function],
            "placeholder": "",
            "title": "",
            "type": "",
            "value": "LyndonOhhyungChun",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": false,
            "isVisible": false,
            "message": "",
            "name": "password",
            "onChange": [Function],
            "placeholder": "",
            "title": "",
            "type": "",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "avatar": Object {
              "message": "",
              "value": "https://lookaside.facebook.com/platform/profilepic/?asid=1527248567354389&height=50&width=50&ext=1522858773&hash=AeTOk-4ueDQXWd1T",
            },
            "className": "User__form__fields",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
